### ISSUE TITLE
Is multiclass segmentation possible?

### ISSUE THREAD

**Issue by dev-walter** [nibtehaz/MultiResUNet#19]
Hello,

I want to implement your model in a cell segmentation task. Since I have to segment my images into 4 classes (background, healthy cells, cancer cells and leucocytes) I wanted to ask if your model can be easily modified to support multiclass segmenation.
I have little experience in deep learning and python so help would be much appreciated!

**Comment by kabbas570** [comment]
Hello, 
yes, you can change, add the following line, at the end of the model's last convolutional layer,
conv10 = conv2d_bn(mresblock9, **4,** 1, 1, activation='sigmoid')
in the original model, this is **1** (number of classes), just change it to **4**, your model can now predict four classes.

**Comment by nibtehaz** [comment]
Hi @dev-walter , thanks for your interest in our project.

Thank you @kabbas570  for resolving the issue.

**Comment by dev-walter** [comment]
> 
> 
> Hello,
> yes, you can change, add the following line, at the end of the model's last convolutional layer,
> conv10 = conv2d_bn(mresblock9, **4,** 1, 1, activation='sigmoid')
> in the original model, this is **1** (number of classes), just change it to **4**, your model can now predict four classes.

Thank you this worked perfectly and the model performed very well :).

**Comment by nibtehaz** [comment]
Really glad to hear that @dev-walter , best wishes for you

**Comment by saskra** [comment]
> I have to segment my images into 4 classes (background, healthy cells, cancer cells and leucocytes)

> add the following line, at the end of the model's last convolutional layer, conv10 = conv2d_bn(mresblock9, **4,** 1, 1, activation='sigmoid') in the original model, this is **1** (number of classes), just change it to **4**, your model can now predict four classes.

Thank you very much, I will test that too. But is the change from 1 to 4 really correct? There was already a background before, so there have always been 2 classes. [Edit: I think I understand it now. With one class plus background, you get the probability of it being the other class, sort of inclusive. Starting from two classes plus background you have to output it explicitly.]

Are there actually any other changes that have become mandatory, e.g. is the Jaccard index in its current form defined for multiclass predictions?

**Comment by nibtehaz** [comment]
Thanks @saskra . Yes for 2 class the background is just the opposite prediction probability. Nevertheless, you can work with 2 class using 2 output classes, but that’s kind of redundant.


in regards to evolution metrics, you can either collapse them to one dimension or class or compute them individually.

hope this helps, best wishes.

**Comment by hendriramdan** [comment]
Hello, sir. 
i'm trying to predict multi class.
in my case, there are 3 classes.
i have changed the number 1 in con_10 to 3 according to the class i have. but i get an error like below:
Logits and labels must have the same shape ((1, 192, 256, 3) vs (1, 192, 256, 1)).
can anyone help me? which part should be changed? 
thanks

**Comment by kabbas570** [comment]
Hi,
change to this..line 192.
   conv10 = conv2d_bn(mresblock9, num_calsses=3, 1, 1, activation='sigmoid')

**Comment by kabbas570** [comment]
If you already did this, then check your target shape.

**Comment by hendriramdan** [comment]
thanks for the response.
I've changed the line to conv10 = conv2d_bn(mresblock9, 3, 1, 1,
activation='sigmoid').
however I get an error : logit and label must have the same form ((1, 192,
256, 3) vs (1, 192, 256, 1)).

to check the target shape in what line?

Thank you

Pada tanggal Sab, 16 Jul 2022 pukul 02.40 Abbas Khan <
***@***.***> menulis:

**Comment by kabbas570** [comment]
check your target shape it can be anywhere according to your implementation. for the code demo check this,
print(Y_train.shape)

**Comment by hendriramdan** [comment]
Y_train =
Y_train.reshape((Y_train.shape[0],Y_train.shape[1],Y_train.shape[2],1))
Y_test = Y_test.reshape((Y_test.shape[0],Y_test.shape[1],Y_test.shape[2],1))

if i change number 1 to number 3. i get error : cannot reshape array of
size 2752512 into shape (56,192,256.3)

Pada tanggal Sab, 16 Jul 2022 pukul 02.58 Hendri Ramdhan <
***@***.***> menulis:

**Comment by kabbas570** [comment]
There is some problem with your ground truth data, then, check it before feeding it to the model.

**Comment by hendriramdan** [comment]
I made ground truth data like this, is there something wrong with my method?
![image](https://user-images.githubusercontent.com/43942980/179303715-5fac60e6-f8f1-41b5-adde-b09acdcec88e.png)

**Comment by kabbas570** [comment]
![image](https://user-images.githubusercontent.com/56618776/179303766-269d0f6b-c7ac-4a52-9a58-f402aa17e489.png)

**Comment by kabbas570** [comment]
arrange your ground truth like this, I hope it will help.

**Comment by hendriramdan** [comment]
I apologize.
when am I exemplified with the original image data?

because in my case, there will divide in 1 image by 3 classes.
is this possible?
thank you very much.

**Comment by hendriramdan** [comment]
if i print Y_train.shape it will show (56, 192, 256, 1).
is my method is correct.
I beg your help sir. because I'm still a beginner.


Pada tanggal Sab, 16 Jul 2022 pukul 02.53 Abbas Khan <
***@***.***> menulis:

**Comment by nibtehaz** [comment]
Hi Hendri, if you want to generate 3 class label, I would suggest you to make the Y_train.shape (56,192,256,3) (or (56,192,256,4) if you wish to have background)
